subsys/bluetooth/host/smp.c

```diff
@@ -3389,64 +3420,90 @@ static uint8_t compute_and_check_and_send_slave_dhcheck(struct bt_smp *smp)
 }
 #endif /* CONFIG_BT_PERIPHERAL */
 
-static void bt_smp_dhkey_ready(const uint8_t *dhkey)
+static void bt_smp_dhkey_ready(const uint8_t *dhkey);
+static uint8_t smp_dhkey_next(struct bt_smp *smp)
 {
-	struct bt_smp *smp = NULL;
-	int i;
-
-	BT_DBG("%p", dhkey);
+	int err;
 
-	for (i = 0; i < ARRAY_SIZE(bt_smp_pool); i++) {
-		if (atomic_test_and_clear_bit(bt_smp_pool[i].flags,
-					      SMP_FLAG_DHKEY_PENDING)) {
-			smp = &bt_smp_pool[i];
-			break;
-		}
+	err = bt_dh_key_gen(smp->pkey, bt_smp_dhkey_ready);
+	if (err) {
+		BT_ERR("Failed to generate DHKey");
+		return BT_SMP_ERR_UNSPECIFIED;
 	}
 
-	if (!smp) {
-		return;
-	}
+	atomic_set_bit(smp->flags, SMP_FLAG_DHKEY_GEN);
```

Comment:
`bt_dh_key_gen` sends the hci command synchronously. Would it be possible for the key_gen to call the ready callback before returning? Either with the current HCI behavior or after any future optimizations to the controller? Perhaps it's better to set this flag before calling key_gen, then unset if there's an error?

It might also be safer to move the `SMP_FLAG_DHKEY_GEN` check in `bt_smp_dhkey_ready` into this function, as it's important that it's checked before we call dh_key_gen to avoid extra errors

Reply:
Since I think that could lead to the opposite problem, dhkey callback called and could match with the wrong smp context.
That really depends on the implementation of the `bt_dh_key_gen`, to avoid depending on that implementation I've thought it would be best to keep this internal in the module, and instead check if the flag has been set in any of smp contexts before calling out of the SMP module.